l1-contracts/contracts/bridgehub/MessageRoot.sol

```diff
@@ -150,12 +307,26 @@ contract MessageRoot is IMessageRoot, Initializable {
         uint256 _chainId,
         uint256 _batchNumber,
         bytes32 _chainBatchRoot
-    ) external onlyChain(_chainId) onlyL2Chain {
+    ) external addChainBatchRootRestriction(_chainId) {
         // Make sure that chain is registered.
         if (!chainRegistered(_chainId)) {
             revert MessageRootNotRegistered();
         }
+        require(_chainBatchRoot != bytes32(0), ChainBatchRootZero());
+        require(
+            chainBatchRoots[_chainId][_batchNumber] == bytes32(0),
+            ChainBatchRootAlreadyExists(_chainId, _batchNumber)
+        );
+        if (_batchNumber > 0) {
+            bytes32 previousBatchNumber = chainBatchRoots[_chainId][_batchNumber - 1];
+            require(previousBatchNumber != bytes32(0), PreviousChainBatchRootNotSet(_chainId, _batchNumber - 1));
```

Comment:
we never submitted any batches on L1, this require wont work.

I think it would be easier for everyone if we stored the `lastExecutedBatchNumber` for each chain and ensured that it is advanced with each `addChainBatchRoot`. 

This way it would also be easier to ensure an additional invariant that the v30 upgrade batch is == lastExecutedBatchNumber at the time of the upgrade, so there are no "jumps", where a random v30 batch number is provided

Reply:
<img width="956" height="235" alt="image" src="https://github.com/user-attachments/assets/a86ecca8-3d04-46db-bcfa-f87830f98615" />
I think it would, because of this. When migrating we need to set some value anyway, either the lastExecutedBatchNumber, or this value.

I'm not sure we need the lastExecutedBatchNumber mapping.

Reply:
Then we call the saveV30UpgradeChainBatchNumber in L1V30upgrade